flink-core/src/main/java/org/apache/flink/core/memory/HybridMemorySegment.java

```diff
@@ -131,6 +131,9 @@
 
     @Override
     public void free() {
+        if (isFreed()) {
+            throw new IllegalStateException("HybridMemorySegment can be freed only once!");
+        }
```

Comment:
I'd suggest to scope this change out of this PR.

I'm afraid current there are other causes that can lead to a segment being freed multiple times. That's also what the CI failures suggest.

As discussed [here](https://issues.apache.org/jira/browse/FLINK-21419?focusedCommentId=17301350&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17301350), we are planning to enable this check-and-fail only for CI and gradually hunt down all the misuse case.

Reply:
Honestly, I was afraid that's gonna be the case that tests would fail, I wanted to check anyway.

The problem without the check though, is that the tests I added pass even without the fix :( Not sure, how can I test the change as I don't have anything really to assert on. Do you have an idea? Or shall we simply leave the tests as they are now and they will make sense once we introduce the check?

Reply:
I'm planning to introduce an environment variable for activating the check. Maybe it makes sense to take that step now, so we can activate the check only for the relevant test cases, before activating it for all tests by default. WDYT?

Reply:
Please check this branch.
https://github.com/xintongsong/flink/tree/flink-21728

Reply:
Thanks! I will incorporate it, once I am done with my meetings.